core/src/main/scala/com/gu/tableversions/core/model.scala

```diff
@@ -1,8 +1,73 @@
 package com.gu.tableversions.core
 
+import java.net.URI
+
+/**
+  * A Partition represents a concrete partition of a table, i.e. a partition column with a specific value.
+  */
+final case class Partition(columnValues: List[Partition.ColumnValue]) {
+
+  /** Given a base path for the table, return the path to the partition. */
+  def resolvePath(tableLocation: URI): URI = {
+    def normalised(dir: URI): URI = if (dir.toString.endsWith("/")) dir else new URI(dir.toString + "/")
+    if (this == Partition.snapshotPartition) {
+      tableLocation
+    } else {
+      val partitionsSuffix =
+        columnValues.map(columnValue => s"${columnValue.column.name}=${columnValue.value}").mkString("", "/", "/")
+      normalised(tableLocation).resolve(partitionsSuffix)
+    }
+  }
+}
+
+object Partition {
+
+  /** Convenience constructor for single column partitions. */
+  def apply(columnValue: ColumnValue): Partition = Partition(List(columnValue))
+
+  /** Convenience constructor for single column partitions. */
+  def apply(column: PartitionColumn, value: String): Partition = Partition(List(ColumnValue(column, value)))
+
+  case class PartitionColumn(name: String) extends AnyVal
+
+  case class ColumnValue(column: PartitionColumn, value: String)
+
+  // The special case partition that represents the root partition of a snapshot table.
+  val snapshotPartition: Partition = Partition(Nil)
+
+}
+
+/**
+  * A partition schema describes the fields used for partitions of a table
+  */
+final case class PartitionSchema(columns: List[Partition.PartitionColumn])
+
+object PartitionSchema {
+
+  // The special case partition that represents the root partition of a snapshot table.
+  val snapshot: PartitionSchema = PartitionSchema(Nil)
+
+}
+
+//
+// Versions
+//
+
+final case class VersionNumber(number: Int) extends AnyVal
+
+final case class PartitionVersion(partition: Partition, version: VersionNumber)
+
 //
-// A partition schema describes the fields used for partitions of a table
+// Tables
 //
-case class PartitionColumn(name: String) extends AnyVal
 
-case class PartitionSchema(keys: List[PartitionColumn])
+final case class TableName(schema: String, name: String) {
```

Comment:
Should we use the one from `tableMetadata`?

Reply:
`tableMetadata` lives in the ophan-data-lake project, and we don't want this project to depend on that.

Reply:
duhhh, of course 😳

Reply:
Also, I had a look at that to see if I could, ahem, borrow the code for it. But, that's a bit different; the storage path for the table lives in the same class as the name and schema (in `TableReference`), and the schema is fixed to a hard-coded list of schema names, so it doesn't really fit here.